### ISSUE TITLE
Int64 with gather results in MLIR-TRT error

### ISSUE THREAD

**Issue by Mgluhovskoi** [NVIDIA/TensorRT-Incubator#115]
Example of what is going wrong:

```
x_shape, axis, indices = ((3, 3), 0, (0,1))
x = np.arange(np.prod(x_shape)).reshape(x_shape)
a = tp.Tensor(x)
indices_tp = tp.Tensor(indices)
a = tp.Tensor(x)
# a = tp.cast(a, dtype=tp.int32)
out = tp.gather(a, axis, indices_tp)
print(out) 
```
Results in the following error:
```
MTRTException: InternalError: failed to run compilation on module with symbol name: outs_t3_1

Additional context:
Traceback (most recent call last):
  File "/tripy/tripy/backend/mlir/compiler.py", line 100, in compile
    executable = compiler.compiler_stablehlo_to_executable(
mlir_tensorrt.runtime._mlir_libs._api.MTRTException: InternalError: failed to run compilation on module with symbol name: outs_t3_1
.
    Loaded TensorRT version 10.1.0.27 but compiled for TensorRT 10.2.0.19. This can result in crashes or unintended behavior.
    (t3)error: op: %2 = "stablehlo.gather"(%0, %1) <{dimension_numbers = #stablehlo.gather<offset_dims = [1], collapsed_slice_dims = [0], start_index_map = [0], index_vector_dim = 1>, slice_sizes = array<i64: 1, 3>}> : (tensor<3x3xi64>, tensor<2xi32>) -> tensor<2x3xi64> from function main is invalid, post clustering.

    This error occured while trying to compile the following FlatIR expression:
          |
          | t3: [rank=(2), shape=((2, -1)), dtype=(int64), loc=(gpu:0)] = DynamicGatherOp(t2, t1, t_inter3, axis=0)
          |
```
If you cast a to int32 before running gather then it executes as expected. It also passes if you change the indices to `tp.Tensor((1))`, `tp.Tensor((0))`, or some other single value. On the other hand `tp.Tensor([0])` and `tp.Tensor([1])` raises the same error as before. Here are the logs:

With indices=`tp.Tensor((0))`: This example passes

```
module @outs_t3_1 {
  func.func @main() -> tensor<?xi64> {
    %c = stablehlo.constant dense<[[0, 1, 2], [3, 4, 5], [6, 7, 8]]> : tensor<3x3xi64>
    %c_0 = stablehlo.constant dense<0> : tensor<i32>
    %c_1 = stablehlo.constant dense<1> : tensor<1xi32>
    %0 = stablehlo.get_dimension_size %c, dim = 0 : (tensor<3x3xi64>) -> tensor<i32>
    %c_2 = stablehlo.constant dense<1> : tensor<1xi32>
    %1 = stablehlo.dynamic_reshape %0, %c_2 : (tensor<i32>, tensor<1xi32>) -> tensor<1xi32>
    %2 = stablehlo.get_dimension_size %c, dim = 1 : (tensor<3x3xi64>) -> tensor<i32>
    %c_3 = stablehlo.constant dense<1> : tensor<1xi32>
    %3 = stablehlo.dynamic_reshape %2, %c_3 : (tensor<i32>, tensor<1xi32>) -> tensor<1xi32>
    %4 = stablehlo.concatenate %1, %3, dim = 0 : (tensor<1xi32>, tensor<1xi32>) -> tensor<2xi32>
    %c_4 = stablehlo.constant dense<1> : tensor<1xi32>
    %c_5 = stablehlo.constant dense<2> : tensor<1xi32>
    %5 = stablehlo.real_dynamic_slice %4, %c_4, %c_5, %c_1 : (tensor<2xi32>, tensor<1xi32>, tensor<1xi32>, tensor<1xi32>) -> tensor<?xi32>
    %6 = stablehlo.concatenate %c_1, %5, dim = 0 : (tensor<1xi32>, tensor<?xi32>) -> tensor<2xi32>
    %7 = "stablehlo.dynamic_gather"(%c, %c_0, %6) <{dimension_numbers = #stablehlo.gather<offset_dims = [0], collapsed_slice_dims = [0], start_index_map = [0]>}> : (tensor<3x3xi64>, tensor<i32>, tensor<2xi32>) -> tensor<?xi64>
    return %7 : tensor<?xi64>
  }
}
``` 

With indices=`tp.Tensor([0]): This example fails

```
module @outs_t3_1 {
  func.func @main() -> tensor<1x?xi64> {
    %c = stablehlo.constant dense<[[0, 1, 2], [3, 4, 5], [6, 7, 8]]> : tensor<3x3xi64>
    %c_0 = stablehlo.constant dense<0> : tensor<1xi32>
    %c_1 = stablehlo.constant dense<1> : tensor<1xi32>
    %0 = stablehlo.get_dimension_size %c, dim = 0 : (tensor<3x3xi64>) -> tensor<i32>
    %c_2 = stablehlo.constant dense<1> : tensor<1xi32>
    %1 = stablehlo.dynamic_reshape %0, %c_2 : (tensor<i32>, tensor<1xi32>) -> tensor<1xi32>
    %2 = stablehlo.get_dimension_size %c, dim = 1 : (tensor<3x3xi64>) -> tensor<i32>
    %c_3 = stablehlo.constant dense<1> : tensor<1xi32>
    %3 = stablehlo.dynamic_reshape %2, %c_3 : (tensor<i32>, tensor<1xi32>) -> tensor<1xi32>
    %4 = stablehlo.concatenate %1, %3, dim = 0 : (tensor<1xi32>, tensor<1xi32>) -> tensor<2xi32>
    %c_4 = stablehlo.constant dense<1> : tensor<1xi32>
    %c_5 = stablehlo.constant dense<2> : tensor<1xi32>
    %5 = stablehlo.real_dynamic_slice %4, %c_4, %c_5, %c_1 : (tensor<2xi32>, tensor<1xi32>, tensor<1xi32>, tensor<1xi32>) -> tensor<?xi32>
    %6 = stablehlo.concatenate %c_1, %5, dim = 0 : (tensor<1xi32>, tensor<?xi32>) -> tensor<2xi32>
    %7 = "stablehlo.dynamic_gather"(%c, %c_0, %6) <{dimension_numbers = #stablehlo.gather<offset_dims = [1], collapsed_slice_dims = [0], start_index_map = [0], index_vector_dim = 1>}> : (tensor<3x3xi64>, tensor<1xi32>, tensor<2xi32>) -> tensor<1x?xi64>
    return %7 : tensor<1x?xi64>
  }
}
``` 

With indices=`tp.Tensor((0,1)): This example fails

```
module @outs_t3_1 {
  func.func @main() -> tensor<2x?xi64> {
    %c = stablehlo.constant dense<[[0, 1, 2], [3, 4, 5], [6, 7, 8]]> : tensor<3x3xi64>
    %c_0 = stablehlo.constant dense<[0, 1]> : tensor<2xi32>
    %c_1 = stablehlo.constant dense<1> : tensor<1xi32>
    %0 = stablehlo.get_dimension_size %c, dim = 0 : (tensor<3x3xi64>) -> tensor<i32>
    %c_2 = stablehlo.constant dense<1> : tensor<1xi32>
    %1 = stablehlo.dynamic_reshape %0, %c_2 : (tensor<i32>, tensor<1xi32>) -> tensor<1xi32>
    %2 = stablehlo.get_dimension_size %c, dim = 1 : (tensor<3x3xi64>) -> tensor<i32>
    %c_3 = stablehlo.constant dense<1> : tensor<1xi32>
    %3 = stablehlo.dynamic_reshape %2, %c_3 : (tensor<i32>, tensor<1xi32>) -> tensor<1xi32>
    %4 = stablehlo.concatenate %1, %3, dim = 0 : (tensor<1xi32>, tensor<1xi32>) -> tensor<2xi32>
    %c_4 = stablehlo.constant dense<1> : tensor<1xi32>
    %c_5 = stablehlo.constant dense<2> : tensor<1xi32>
    %5 = stablehlo.real_dynamic_slice %4, %c_4, %c_5, %c_1 : (tensor<2xi32>, tensor<1xi32>, tensor<1xi32>, tensor<1xi32>) -> tensor<?xi32>
    %6 = stablehlo.concatenate %c_1, %5, dim = 0 : (tensor<1xi32>, tensor<?xi32>) -> tensor<2xi32>
    %7 = "stablehlo.dynamic_gather"(%c, %c_0, %6) <{dimension_numbers = #stablehlo.gather<offset_dims = [1], collapsed_slice_dims = [0], start_index_map = [0], index_vector_dim = 1>}> : (tensor<3x3xi64>, tensor<2xi32>, tensor<2xi32>) -> tensor<2x?xi64>
    return %7 : tensor<2x?xi64>
  }
}
```

With indices=`tp.Tensor((0,1)) and casting input tensor to int32: This example passes

```
module @outs_t4_1 {
  func.func @main() -> tensor<2x?xi32> {
    %c = stablehlo.constant dense<[[0, 1, 2], [3, 4, 5], [6, 7, 8]]> : tensor<3x3xi64>
    %0 = stablehlo.convert %c : (tensor<3x3xi64>) -> tensor<?x?xi32>
    %c_0 = stablehlo.constant dense<[0, 1]> : tensor<2xi32>
    %c_1 = stablehlo.constant dense<1> : tensor<1xi32>
    %1 = stablehlo.get_dimension_size %0, dim = 0 : (tensor<?x?xi32>) -> tensor<i32>
    %c_2 = stablehlo.constant dense<1> : tensor<1xi32>
    %2 = stablehlo.dynamic_reshape %1, %c_2 : (tensor<i32>, tensor<1xi32>) -> tensor<1xi32>
    %3 = stablehlo.get_dimension_size %0, dim = 1 : (tensor<?x?xi32>) -> tensor<i32>
    %c_3 = stablehlo.constant dense<1> : tensor<1xi32>
    %4 = stablehlo.dynamic_reshape %3, %c_3 : (tensor<i32>, tensor<1xi32>) -> tensor<1xi32>
    %5 = stablehlo.concatenate %2, %4, dim = 0 : (tensor<1xi32>, tensor<1xi32>) -> tensor<2xi32>
    %c_4 = stablehlo.constant dense<1> : tensor<1xi32>
    %c_5 = stablehlo.constant dense<2> : tensor<1xi32>
    %6 = stablehlo.real_dynamic_slice %5, %c_4, %c_5, %c_1 : (tensor<2xi32>, tensor<1xi32>, tensor<1xi32>, tensor<1xi32>) -> tensor<?xi32>
    %7 = stablehlo.concatenate %c_1, %6, dim = 0 : (tensor<1xi32>, tensor<?xi32>) -> tensor<2xi32>
    %8 = "stablehlo.dynamic_gather"(%0, %c_0, %7) <{dimension_numbers = #stablehlo.gather<offset_dims = [1], collapsed_slice_dims = [0], start_index_map = [0], index_vector_dim = 1>}> : (tensor<?x?xi32>, tensor<2xi32>, tensor<2xi32>) -> tensor<2x?xi32>
    return %8 : tensor<2x?xi32>
  }
}
```

With indices=`tp.Tensor([0])` and casting input tensor to int32: This example passes

```
module @outs_t4_1 {
  func.func @main() -> tensor<1x?xi32> {
    %c = stablehlo.constant dense<[[0, 1, 2], [3, 4, 5], [6, 7, 8]]> : tensor<3x3xi64>
    %0 = stablehlo.convert %c : (tensor<3x3xi64>) -> tensor<?x?xi32>
    %c_0 = stablehlo.constant dense<0> : tensor<1xi32>
    %c_1 = stablehlo.constant dense<1> : tensor<1xi32>
    %1 = stablehlo.get_dimension_size %0, dim = 0 : (tensor<?x?xi32>) -> tensor<i32>
    %c_2 = stablehlo.constant dense<1> : tensor<1xi32>
    %2 = stablehlo.dynamic_reshape %1, %c_2 : (tensor<i32>, tensor<1xi32>) -> tensor<1xi32>
    %3 = stablehlo.get_dimension_size %0, dim = 1 : (tensor<?x?xi32>) -> tensor<i32>
    %c_3 = stablehlo.constant dense<1> : tensor<1xi32>
    %4 = stablehlo.dynamic_reshape %3, %c_3 : (tensor<i32>, tensor<1xi32>) -> tensor<1xi32>
    %5 = stablehlo.concatenate %2, %4, dim = 0 : (tensor<1xi32>, tensor<1xi32>) -> tensor<2xi32>
    %c_4 = stablehlo.constant dense<1> : tensor<1xi32>
    %c_5 = stablehlo.constant dense<2> : tensor<1xi32>
    %6 = stablehlo.real_dynamic_slice %5, %c_4, %c_5, %c_1 : (tensor<2xi32>, tensor<1xi32>, tensor<1xi32>, tensor<1xi32>) -> tensor<?xi32>
    %7 = stablehlo.concatenate %c_1, %6, dim = 0 : (tensor<1xi32>, tensor<?xi32>) -> tensor<2xi32>
    %8 = "stablehlo.dynamic_gather"(%0, %c_0, %7) <{dimension_numbers = #stablehlo.gather<offset_dims = [1], collapsed_slice_dims = [0], start_index_map = [0], index_vector_dim = 1>}> : (tensor<?x?xi32>, tensor<1xi32>, tensor<2xi32>) -> tensor<1x?xi32>
    return %8 : tensor<1x?xi32>
  }
}
``` 
Please let me know if any more information is needed.

**Comment by christopherbate** [comment]
The TensorRT dialect doesn't have coverage of i64 yet. Right now it's not a priority. If you need it, then feel free to expand I64 support in [the TensorRT operator ODS spec](https://github.com/NVIDIA/TensorRT-Incubator/issues/tensorrt/include/mlir-tensorrt-dialect/TensorRT/IR/TensorRTOps.td) everywhere that is relevant.

Once that is done, we would need to update StableHlo-to-TensorRT. @shelkesagar29 this is another area where we need appropriate use of the target TensorRT version in the conversion pass. I64 should only be allowed as a target datatype for the TensorRT versions that support I64.

**Comment by christopherbate** [comment]
Duplicate of https://github.com/NVIDIA/TensorRT-Incubator/issues/116 (i64 support in TensorRT dialect)
